### PR TITLE
Add parentheses properly for borrowing suggestion

### LINKED ISSUE #104961: "consider borrowing here" suggestion is missing parentheses
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c03c22d0ff9f0d88bb0cacb6edb95501

```rust
fn foo(x: &str) -> bool {
    x.starts_with("hi".to_string() + " you")
}
```

The current output is:

```
error[[E0277]](https://doc.rust-lang.org/stable/error-index.html#E0277): expected a `FnMut<(char,)>` closure, found `String`
 --> src/lib.rs:2:19
  |
2 |     x.starts_with("hi".to_string() + " you")
  |       ----------- ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Pattern<'_>` is not implemented for `String`
  |       |
  |       required by a bound introduced by this call
  |
  = note: the trait bound `String: Pattern<'_>` is not satisfied
  = note: required for `String` to implement `Pattern<'_>`
note: required by a bound in `core::str::<impl str>::starts_with`
help: consider borrowing here
  |
2 |     x.starts_with(&"hi".to_string() + " you")
  |                   +
```

Ideally the output should look like:

```
error[[E0277]](https://doc.rust-lang.org/stable/error-index.html#E0277): expected a `FnMut<(char,)>` closure, found `String`
 --> src/lib.rs:2:19
  |
2 |     x.starts_with("hi".to_string() + " you")
  |       ----------- ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Pattern<'_>` is not implemented for `String`
  |       |
  |       required by a bound introduced by this call
  |
  = note: the trait bound `String: Pattern<'_>` is not satisfied
  = note: required for `String` to implement `Pattern<'_>`
note: required by a bound in `core::str::<impl str>::starts_with`
help: consider borrowing here
  |
2 |     x.starts_with(&("hi".to_string() + " you"))
  |                   ++
```

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #104961

### CODE REVIEW

**Comment by chenyukang** [comment]
It's a hacky fix, but seems we also use `snippet` for this.
There should be a completed fix if we can get `expr` kind stuff here, like this:
https://github.com/rust-lang/rust/blob/df04d28163cc42e2c5b1072abe68460f904e42be/compiler/rustc_hir_typeck/src/demand.rs#L830

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1118,13 +1118,22 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         );
                     } else {
                         let is_mut = mut_ref_self_ty_satisfies_pred || ref_inner_ty_mut;
+                        // FIXME: hacky way to get the correct suggestion, see #104961
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1121]: Yes, let me check how we passing <expr> to here, I had a dig yesterday, seems need lot of changes.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -1351,13 +1352,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         );
                     } else {
                         let is_mut = mut_ref_self_ty_satisfies_pred || ref_inner_ty_mut;
+
+                        // see #104961
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1356]: Could you elaborate the comment? Links to issues make understanding harder than a sentence describing the issue.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -1351,13 +1352,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+                            span,
                             &format!(
                                 "consider{} borrowing here",
                                 if is_mut { " mutably" } else { "" }
                             ),
-                            format!("&{}", if is_mut { "mut " } else { "" }),
+                            if !needs_parens {
+                                sugg_prefix
+                            } else {
+                                format!("{}({})", sugg_prefix, snippet)
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1381]: Suggestions that copy user code are typically hard to understand: everything is underlined, and the difference is not underlined.
Could you prefer a multipart suggestion with `&(` at the beginning and `)` at the end?

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 4]
```diff
@@ -1351,13 +1352,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         );
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1352]: Should the parentheses be added to this suggestion too?

Comment (chenyukang) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1352]: I don't think so, I go through all the UI testcase come from this part,
they are all with single argument, and scenrio here is both `&` and `&mut` can be added for borrowing.

I haven't figure out an case which we should add parentheses here.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: make sense, fixed.

**Pull request merged**